### PR TITLE
Introduce hard span storage size limit in sweeper

### PR DESCRIPTION
**Pull request by deadtrickster**
Let's discuss while I'm thinking about tests

### CODE REVIEW

File: src/oc_span_sweeper.erl [hunk 1]
```diff
@@ -88,6 +77,40 @@ terminate(_Reason, _State, _Data) ->
 
 %%
 
+do_gc(#data{strategy=Strategy,
+            ttl=TTL,
+            storage_size=MaxSize}) ->
+
+    StorageSize = ets:info(?SPAN_TAB, memory) * erlang:system_info({wordsize, external}),
```
Comment (tsloughter) [inline src/oc_span_sweeper.erl]: Shouldn't bother with these when `storage_size` is `infinity`.

**Comment by tsloughter** [comment]
I'm not opposed to this, but I do think something like this makes more sense to effect the sampling. Like a sampler that lowers the probability relative to the size of the table.

File: src/oc_span_sweeper.erl (hunk 1 above)
Comment (deadtrickster) [reply src/oc_span_sweeper.erl]: oh right

**Comment by deadtrickster** [comment]
well, 'overflow' could happen probably with any sampler other than none. So to me it's a bit different thing, like literally what shields from crashes. Looking forward to more fancy samples though

**Comment by tsloughter** [comment]
Sure, and the sampler that lowers based on size could go all the way to `none` in times over overload.

**Comment by deadtrickster** [comment]
but it means that only this particular sampler cares about table size, not all of them? I mean if this code stays here all samplers are covered, otherwise I have to stick to one sampler

File: src/oc_span_sweeper.erl [hunk 2]
```diff
@@ -58,24 +61,10 @@ handle_event(state_timeout, sweep, _, #data{interval=Interval,
...
-                                            ttl=TTL}) ->
-    Expired = select_expired(TTL),
-    [finish_span(oc_span:put_attribute(<<"finished_by_sweeper">>, true, Span)) || Span <- Expired],
-    {keep_state_and_data, [hibernate, {state_timeout, Interval, sweep}]};
-handle_event(state_timeout, sweep, _, #data{interval=Interval,
-                                            strategy=Fun,
-                                            ttl=TTL}) when is_function(Fun) ->
-    Expired = select_expired(TTL),
-    [Fun(Span) || Span <- Expired],
+    {keep_state_and_data, [{next_event, internal, just_do_it},
```
Comment (tsloughter) [inline src/oc_span_sweeper.erl]: What is this new internal event?

File: src/oc_span_sweeper.erl [hunk 3]
```diff
@@ -88,6 +77,45 @@ terminate(_Reason, _State, _Data) ->
 
 %%
 
+do_gc(#data{strategy=Strategy,
+            ttl=TTL,
+            storage_size=MaxSize}) ->
+
+    StorageSize = case MaxSize of
+                      infinity ->
```
Comment (tsloughter) [inline src/oc_span_sweeper.erl]: Not a big deal, but personally I'd find it cleaner to be like:

```erlang
do_gc(#data{strategy=Strategy,
            ttl=TTL,
            storage_size=infinity}) ->
    sweep_spans(Strategy, TTL);
do_gc(#data{strategy=Strategy,
            ttl=TTL,
            storage_size=MaxSize}) ->
    StorageSize = ets:info(?SPAN_TAB, memory) * erlang:system_info({wordsize, external}),
    ...
```

File: src/oc_span_sweeper.erl (hunk 2 above)
Comment (deadtrickster) [reply src/oc_span_sweeper.erl]: copy-pasta, at the beginning I wanted to first sweep by ttl and then memory, that's a left over

File: src/oc_span_sweeper.erl (hunk 3 above)
Comment (deadtrickster) [reply src/oc_span_sweeper.erl]: was thinking about this, thought would be ugly, but now I see it's not

**Comment by deadtrickster** [comment]
Elvis complains about `if`. Imo looks rather nice. ignore?

**Comment by tsloughter** [comment]
Weird, yea, I would ignore that from elvis.

**Pull request merged**